liwords-ui/src/chat/players.tsx

```diff
@@ -0,0 +1,294 @@
+    const tabPaneHeight = document.getElementById('chat')?.clientHeight;
+    setMaxHeight(tabPaneHeight ? tabPaneHeight - 117 : undefined);
+  }, []);
+
+  useEffect(() => {
+    setHeight();
+  }, [setHeight]);
+  const onPlayerSearch = useCallback(
+    (searchText: string) => {
+      if (searchText?.length > 0) {
+        axios
+          .post<SearchResponse>(
+            toAPIUrl('user_service.AutocompleteService', 'GetCompletion'),
+            {
+              prefix: searchText,
+            }
+          )
+          .then((resp) => {
+            // Exclude your friends
+            const nonfriends = resp.data.users.filter((u) => {
+              return !Object.values(friends).some(
+                (f) => f.username.toLowerCase() === u.username?.toLowerCase()
+              );
+            });
+            // Exclude yourself
+            setSearchResults(
+              !searchText
+                ? []
+                : nonfriends.filter(
+                    (u) => u.username?.toLowerCase() !== username.toLowerCase()
+                  )
+            );
+          });
+      } else {
+        setSearchResults([]);
+      }
+    },
+    [username, friends]
+  );
+  const searchUsernameDebounced = debounce(onPlayerSearch, 200);
+
+  const handleSearchChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      const prefix = e.target.value;
+      setSearchText(prefix);
+      if (prefix?.length > 0) {
+        searchUsernameDebounced(prefix);
+      }
+    },
+    [searchUsernameDebounced]
+  );
+
+  const renderPlayerList = useCallback(
+    (userList: Partial<FriendUser>[], className: string = ''): ReactNode => {
+      return (
+        <>
+          {userList.map((p) => (
+            <Player
+              sendMessage={sendMessage}
+              className={className}
+              key={p.uuid}
+              {...p}
+            />
+          ))}
+        </>
+      );
+    },
+    [sendMessage]
+  );
+
+  const filterPlayerListBySearch = useCallback(
+    (searchTerm: string, list: Partial<FriendUser>[]) => {
+      if (searchTerm?.length) {
+        return list.filter((u) =>
+          u.username?.toLowerCase().startsWith(searchTerm.toLowerCase())
+        );
+      } else {
+        return list;
+      }
+    },
+    []
+  );
+
+  const transformAndFilterPresences = useCallback(
+    (
+      presenceUsers: PresenceEntity[],
+      searchTerm: string
+    ): Partial<FriendUser>[] => {
+      const presencePlayers = presenceUsers
+        .filter((p) => !p.anon)
+        .sort((a, b) => {
+          if (a.username < b.username) {
+            return -1;
+          }
+          if (a.username > b.username) {
+            return 1;
+          }
+          return 0;
+        })
+        .map(
+          (p): Partial<FriendUser> => {
+            return {
+              username: p.username,
+              uuid: p.uuid,
+              channel: new Array<string>().concat(p.channel),
```

Comment:
this would be `channel: [p.channel],` (turning a string to a string array, without first allocating an empty array; `concat` returns a new array).

but are the items in `PresenceEntity` all in the same channel? or are there potentially going to be duplicates if they are in multiple channels?

Reply:
Nice catch. You get duplicates in club games, since your subscribed to both the club chat and the game or gametv chat. Refactoring.